### ISSUE TITLE
noFallthroughCasesInSwitch complains about last case

### ISSUE THREAD

**Issue by domoritz** [Microsoft/TypeScript#6712]
I was a bit surprised that the last case also requires a break with `noFallthroughCasesInSwitch` in 1.8. Is that intentional? 

```ts
switch (x % 2) {
    case 0:
        console.log("even");
         break;
    case 1: // Error: Fallthrough case in switch.
        console.log("odd");
    // no break here
}
```

**Comment by RyanCavanaugh** [comment]
On the one hand, it's not technically a fall-through.

On the other hand, it's a maintenance trap for the next person who adds a `case` label.

:question: :confused: :question:

**Comment by domoritz** [comment]
This actually is consistent with what tslint does. 

@RyanCavanaugh 
One could argue that if the next person adds another case, ts will throw the error. But there are certainly more important things to do so the design makes sense.

**Comment by fongandrew** [comment]
This is also triggering on the `default` case at the end. Presumably, no one is adding cases after `default`, so that shouldn't count as a fall-through or a maintenance issue, right?

**Comment by mhegazy** [comment]
For consistency, the default label is not treated like any of the other case labels.

**Comment by aleung** [comment]
@domoritz Tslint no-switch-case-fall-through rule doesn't require a break in last branch.

**Comment by domoritz** [comment]
@aleung You're right. I mixed up two things. Tslint does not require a break for the last case.

**Comment by martine** [comment]
As an experiment we attempted enabling this flag for the TypeScript code at Google (~1k source files across multiple teams).

It caught one bug, found one case where the fallthrough was intentional (there was a comment indicating as such), and triggered on ten other cases where the code was already correct and didn't include the final break.

With a 1 in 10 useful triggering rate, it's the sort of thing we might not be able to use, or need to put into some tool that runs offline in some sort of advisory role, which is kind of a bummer.  Does the additional information about this being inconsistent with tslint influence your decision any?

(I think the "maintenance trap" argument Ryan made isn't quite right, because if next person who adds a case label doesn't add the appropriate break, their error will be caught by exactly this compiler check.)

**Comment by RyanCavanaugh** [comment]
I'll bring it up for discussion.

As for your hit rate -- finding 1 bug at the cost of adding 9 mechanical `break;` statements (should take 2 minutes tops?) sounds like an *insanely* good ROI. If I could find bugs at that cost, I'd do it at the exclusion of all other activity.

**Comment by alexeagle** [comment]
Did you guys roll out this check to any MS codebase before finalizing the implementation? You could have a much higher ROI if you detected only the cases where fallthrough has another case to fall into.

I think it's important to define the goals of static analysis in the compiler. Could I join in for the discussion?

(I spent a year making static analysis happen in Google's Java compiler. See http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43322.pdf and we have a paper coming soon specifically about http://ErrorProne.info which I started)

**Comment by martine** [comment]
@RyanCavanaugh That's a good way to look at it -- the implementation cost is really low, so it might make sense to implement it regardless.

But for a bit more context, we brought up adding this check internally and some users seemed like they'd be a bit cranky about it because their code is already "correct".  So if it doesn't cost much, fixing this bug means we'd have had only two hits, one of which was a bug and the other was some code that was kind of sketchy to begin with, which is a pretty awesome success rate.

**Comment by domoritz** [comment]
I found @martine and @RyanCavanaugh perspectives very interesting. The 1/10 rate is good and catching potential errors is always great. However, I disabled the check in our codebase because we had and intended fall though that was marked as such and we couldn't compile the code any more. We use [tslint](https://github.com/palantir/tslint) which catches unintended fall thoughs already but it allows us to ignore certain checks in part of the code but enable them in general. 

More generally, I wonder where typescript draws the line between compile errors, warnings, and linting. I understand that typescript itself may be better at static analysis than an external linting tool but I'd hope to see different rules for the different categories of errors. Compile errors should stop the build, errors like this may stop the build unless explicitly silences for a particular line/case.

**Comment by RyanCavanaugh** [comment]
It's a very good question and one we're having to think even harder about now that TypeScript is powering JavaScript intellisense in some Visual Studio products (which is an even harder scenario because there's not an obvious place for the user to configure their desired behavior). People don't really like it when we tell them "Use TSLint if you want that to be an error" but we don't want to be in the business of managing 600 commandline switches either.

Getting feedback like this is great for helping us validate our mental models of how people write code in the real world.

**Comment by tinganho** [comment]
>People don't really like it when we tell them "Use TSLint if you want that to be an error" but we don't want to be in the business of managing 600 commandline switches either

I think the time between when your error is created to the time you discovered it, is crucial. I think a long duration will be considered as noise and not feedback. Even though it technically is a feedback.

I personally like linting but it could be multiple times better if it could be integrated into the language service. I want to see errors in my editor or IDE as red squiggles, because the duration between creating it, discovering and fixing it, is near zero.

**Comment by domoritz** [comment]
> I think the time between when your error is created to the time you discovered it, is crucial. I think a long duration will be considered as noise and not feedback. Even though it technically is a feedback.

This is an orthogonal issue. When I was talking about errors vs errors that can be silenced (lint) it has nothing to do with when the user get's to see the error. Typescript can compile things in the background and provide immediate inline hints as soon as possible.

**Comment by alexeagle** [comment]
We have specific criteria for when to add errors in the Java compiler:
http://errorprone.info/docs/criteria

Having a bug caught early is indeed important. If you run tslint in your build system or continuous integration, it may be too late to catch the bug you spent 30 minutes debugging, because you don't run the build system in your edit-refresh-debug development cycle. TypeScript compiler is the only component which reliably gets to reject incorrect programs before you manually debug them.

The ability to configure checks can be available both in the compiler and other tools. I think @RyanCavanaugh is right to be concerned about supporting a large configuration surface (which is beginning to happen with the handful of new compiler options in 1.8). But a suppression mechanism can also be allowed in the source code, with a marker comment or Decorator. In Java, we have both source level `@SuppressWarning("StringEquality")` and compiler flag `-Xep:StringEquality:OFF`.

The bigger motivation for us is that warnings are very different from errors, because they may be permitted to accumulate in code. Warning blindness means that developers ignore the warnings when convenient, because there is no guarantee that any of these warnings were just introduced. So we only show warnings in code review, and in fact have disabled all existing Java compiler warnings from appearing in the output.

**Comment by domoritz** [comment]
In our project, we treat lint warnings as errors and don't accept any code that does not conform to our linting rules. Our watch script automatically compile and lint the code with every change (we also have manual build and lint scripts). In addition, we use plugins that integrate typescript and tslint into our editors. This means we get to see errors and warnings as we type. 

> I think @RyanCavanaugh is right to be concerned about supporting a large configuration surface (which is beginning to happen with the handful of new compiler options in 1.8). But a suppression mechanism can also be allowed in the source code, with a marker comment or Decorator.

That sounds a lot like what tslint already does ;-) It seems good to have a discussion about what rules/ which analysis goes into the compiler or some other tool.

**Comment by domoritz** [comment]
The "The bug pattern should have no false positives. In essentially no cases should the detected code actually be working as intended." rule from http://errorprone.info/docs/criteria excludes decorators for compile errors. So `noFallthroughCasesInSwitch` should not be a compile error but instead a warning (which can be made a compile error with a flag and silenced with a decorator).
